modules/sfp_dehashed.py

```diff
@@ -0,0 +1,219 @@
+        'api_key': 'Dehashed API key.',
+        'per_page': 'Maximum number of results per page.(Max: 10000)',
+        'max_pages': 'Maximum number of pages to fetch(Max: 10 pages)',
+        'pause': 'Number of seconds to wait between each API call.'
+    }
+
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.results = self.tempStorage()
+
+        for opt in list(userOpts.keys()):
+            self.opts[opt] = userOpts[opt]
+
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return [
+            "DOMAIN_NAME",
+            "EMAILADDR"
+        ]
+
+    # What events this module produces
+    def producedEvents(self):
+        return [
+            'EMAILADDR',
+            'EMAILADDR_COMPROMISED',
+            'PASSWORD_COMPROMISED',
+            'HASH_COMPROMISED',
+            'RAW_RIR_DATA'
+        ]
+
+    # Query Dehashed
+    def query(self, event, per_page, start):
+        if event.eventType == "EMAILADDR":
+            queryString = f"https://api.dehashed.com/search?query=email:\"{event.data}\"&page={start}&size={self.opts['per_page']}"
+        if event.eventType == "DOMAIN_NAME":
+            queryString = f"https://api.dehashed.com/search?query=email:\"@{event.data}\"&page={start}&size={self.opts['per_page']}"
+
+        token = (base64.b64encode(self.opts['api_key_username'].encode('utf8') + ":".encode('utf-8') + self.opts['api_key'].encode('utf-8'))).decode('utf-8')
+        headers = {
+            'Accept': 'application/json',
+            'Authorization': f'Basic {token}'
+        }
+
+        res = self.sf.fetchUrl(queryString,
+                               headers=headers,
+                               timeout=15,
+                               useragent=self.opts['_useragent'],
+                               verify=True)
+
+        time.sleep(self.opts['pause'])
+
+        if res['code'] == "400":
+            self.sf.error("Too many requests were performed in a small amount of time. Please wait a bit before querying the API.")
+            time.sleep(5)
+            res = self.sf.fetchUrl(queryString, headers=headers, timeout=15, useragent=self.opts['_useragent'], verify=True)
+
+        if res['code'] == "401":
+            self.sf.error("Invalid API credentials")
+            self.errorState = True
+            return None
+
+        if res['code'] != "200":
+            self.sf.error("Unable to fetch data from Dehashed.")
+            self.errorState = True
+            return None
+
+        if res['content'] is None:
+            self.sf.debug('No response from Dehashed')
+            return None
+
+        try:
+            return json.loads(res['content'])
+        except Exception as e:
+            self.sf.debug(f"Error processing JSON response: {e}")
+            return None
+
+    # Handle events sent to this module
+    def handleEvent(self, event):
```

Comment:
This needs to prevent reporting duplicate events. If you test the module with airbnb.com, you'll see many of the same emails and breaches reported multiple times.

Reply:
Update : 
-Skipping same email + breached data
- not reporting same emails